### DISCUSSION TITLE
Run plugins before Maven-Tycho build

### DISCUSSION THREAD

**Discussion by LuluDavid** [eclipse-tycho/tycho#434] [Q&A]
Hi, I am creating an issue because I am working currently with maven-tycho-plugin for a bunch of OSGI bundles, and I have a small problem to include preliminary instructions in my maven build before tycho's dependency resolution.  
The problem is : 
- Tycho (or potentially plexus) ties its resolution to the validate phase (as described in [this file](https://github.com/eclipse/tycho/blob/master/tycho-maven-plugin/src/main/resources/META-INF/plexus/components.xml), which is the first executed phase
- I am packaging my bundles as eclipse-plugin (so the resolution is done for all of them)
- All further plugins executions tied to the validate phase are executed afterwards (in accordance to Maven's lifecycle description)  

So in my case, as I want to download a jar that is referenced locally in one of my bundle's classpath before Tycho's dependency resolution, it does not work for the first build (but works for the second one as the jar is available then).  

I found [this issue](https://tycho-user.eclipse.narkive.com/9eUeEcjk/running-maven-plugin-before-tycho-packaging-plugin-0-25-0-build-qualifier-default-build-qualifier) that references a similar problem, but the concerned pom is packaged as pom and I tried overriding tycho-maven-plugin's executions to tie all validate-related ids (org.eclipse.tycho:tycho-packaging-plugin:${project.version}:build-qualifier,org.eclipse.tycho:tycho-packaging-plugin:${project.version}:validate-id, org.eclipse.tycho:tycho-packaging-plugin:${project.version}:validate-version) to the initialize phase instead : 
```xml
<pluginManagement>
      ...
      <plugin>
	      <groupId>org.eclipse.tycho</groupId>
	      <artifactId>tycho-maven-plugin</artifactId>
	      <version>${tycho.version}</version>
	      <extensions>true</extensions>
	      <executions>
		      <execution>
			      <id>org.eclipse.tycho:tycho-packaging-plugin:${project.version}:build-qualifier</id>
			      <phase>initialize</phase>
		      </execution>
		      <execution>
			      <id>org.eclipse.tycho:tycho-packaging-plugin:${project.version}:validate-id</id>
			      <phase>initialize</phase>
		      </execution>
		      <execution>
			      <id>org.eclipse.tycho:tycho-packaging-plugin:${project.version}:validate-version</id>
			      <phase>initialize</phase>
		      </execution>
	      </executions>
      </plugin>
</pluginManagement>
```
Without further success.  

If you have any insight or idea to solve this issue, or if my question is not relevant, feel free to answer.
Thanks for reading !

**Comment by laeubi** [comment]
Tycho actually runs as an extension outside the regular life-cycle.

> in my case, as I want to download a jar that is referenced locally in one of my bundle's classpath

classpath resolution should take place in the compile phase, so maybe you can explain a bit more about your "download" and why there is no other way to provide it to your build (e.g. as a maven dependency).

Theoretically it should work to have an (empty) default file in the repo that is later on overwritten ... but this depends a bit on the current error.

**Comment by LuluDavid** [comment]
@laeubi  Thanks for the quick answer.  
Oh ok, I thought it tied its different executions to the existing lifecycle.  
The jar is served on a server on our intranet, and we wanted to include it with a simple wget :
```xml
<profiles>
	<!-- Download jar if missing -->
	<profile>
		<activation>
			<file>
				<missing>${basedir}/lib/${somePlugin.fileName}</missing>
			</file>
		</activation>
		<build>
			<plugins>
				<plugin>
					<groupId>org.codehaus.mojo</groupId>
					<artifactId>exec-maven-plugin</artifactId>
					<version>${exec-maven-plugin.version}</version>
					<executions>
						<execution>
							<id>jar-dl</id>
							<phase>validate</phase>
							<goals>
								<goal>exec</goal>
							</goals>
							<configuration>
								<executable>wget</executable>
								<workingDirectory>lib</workingDirectory>
								<arguments>
									<argument>https://url/to/${somePlugin.fileName}</argument>
								</arguments>
							</configuration>
						</execution>
					</executions>
				</plugin>
			</plugins>
		</build>
	</profile>
</profiles>
```
Maybe my approach is not conventional, in which case I will take any suggestion on the right way to do it.

**Reply by laeubi** [reply]
I think `initialize` would be a more semantic phase to bind here. Can you share the error you are getting when the file is missing?

**Reply by LuluDavid** [reply]
Sure, I think so too (it was the initial configuration but I moved it to validate to produce it at the earliest stage of the build).  
I did not want to over-complicate things, but we are using [spring boot maven plugin](https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/) to produce executable jars, and basically :  
- The code above is in our shared bundle (which contains all local jar dependencies for our bundles)
- We build executable jars that include this shared bundle (and thus use the classes of someJar)
  
And at first build (```mvn clean verify```), when executing tests at the end with one of the executable jars, I get a NoClassDefFoundError for one of the classes of someJar. But if I build again (the jar is now available thanks to the code above), it is packaged as expected in the executable jar.  

Maybe it is spring-boot's plugin that does not resolve the dependency from classpath as it is not downloaded yet, but I got a strong feeling it was an error from Tycho-Maven plugin when I saw the Tycho log entries :
```
[INFO] Resolving dependencies of MavenProject: ...
[INFO] Resolving class path of MavenProject: ...
```
In the output of the build **before** the downloading of the jar.

**Reply by laeubi** [reply]
Your build setup is far from what the usual tycho build looks like so I fear you have to provide at a least a minimal reproducer project to get some help on this topic.

**Comment by Bananeweizen** [comment]
If you can't trigger the download early enough inside the current POM by selecting a phase for it, then just move it to a separate POM and make that new POM a dependency of the current one. That would be the most simple workaround without fully understanding the root cause. dependencies are definitely built before their dependee.
Also if you are working with profiles, the execution order of plugins can become unpredictable (at least if the same plugin is used in multiple profiles). I would therefore replace the profile with curl by an unconditional call to maven-dependency-plugin:get (or unpack or...). Since that has built-in caching, there is no performance penalty in always calling it.

**Comment by laeubi** [comment]
@LuluDavid I think the issue might be related to 
- https://github.com/eclipse/tycho/issues/460

because the classpath is resolve before the regular maven build start your jar is not considered if not present beforehand.

**Comment by LuluDavid** [comment]
@Bananeweizen I was working with profiles at first, but I could not even make it work when I removed them for a simple curl. The problem is that my top level POM is declaring tycho for all modules, and one of these modules needs a preliminary download of an external jar. So I am not sure if I could make it work without breaking the hierarchy of my POMs.  
@laeubi Sorry for not answering before, I could not find the time to produce a minimal example.  
Thanks for referencing this issue here, it seems indeed that it could be related to my issue, I should be able to tell you in a near future (we first need to migrate to java 11 & wait for the release of your fix in a next version of the plugin).

**Reply by laeubi** [reply]
You can use the current [tycho snapshot](https://github.com/eclipse/tycho/wiki#getting-tycho-snapshots) for verification, also if you project uses java 8 you can still use java 11 to run the tycho build.

**Reply by Bananeweizen** [reply]
> The problem is that my top level POM is declaring tycho for all modules, and one of these modules needs a preliminary download of an external jar.

Doesn't matter. Create a new POM, run the download there, and either make that one specific module or the parent depend on the new POM (depending on whether you want the download to happen immediately before that one module, or as the very first thing in the complete aggregator).

**Reply by LuluDavid** [reply]
@laeubi I tried using running the tycho build in java 11 (setting my JAVA_HOME to a jdk 11) for my java 8 project, which led me to enforce the execution environment java 8 for dependency resolution : 
```xml
<plugin>
   <groupId>org.eclipse.tycho</groupId>
   <artifactId>target-platform-configuration</artifactId>
   <version>${tycho.version}</version>
   <configuration>
      <!-- Optional set the Java version your are using -->
      <executionEnvironment>JavaSE-1.8</executionEnvironment>
      ...
   </configuration>
</plugin>
```
But I eventually get an error with my target platform resolution : 
```bash
[ERROR] Cannot resolve target definition:
[ERROR]   Software being installed: epp.package.committers 4.7.3.20180405-1200
[ERROR]   Missing requirement: epp.package.committers 4.7.3.20180405-1200 requires 'org.eclipse.equinox.p2.iu; a.jre.javase [1.6.0,1.6.0]' but it could not be found
[ERROR] 
```
If I do not specify it, it automatically uses java 11, which also fails to resolve target platform with the same error.
And if I set it to none, I get a similar issue than described [here](https://github.com/eclipse/tycho/issues/154).

**Reply by laeubi** [reply]
I think you need to [configure a toolchain.xml](https://wiki.eclipse.org/Tycho/Execution_Environments#Toolchains) where you have a "real" JDK8 in the path. But I'm not using Java 8 for a while now maybe @cdietrich can give some hint I think xtext/xtend has also a java8 tycho build?

**Comment by LuluDavid** [comment]
@Bananeweizen In a similar fashion to your hint, I moved the download to the first module of the build (which makes sense because it is a platform target project), and it worked. So I guess the precise limitation that prevented me to download before tycho's resolution was that I tried to do it in an eclipse-plugin packaged maven project, so your idea to externalize such tasks from the plugin was the solution.  

Anyway, @laeubi's fix should make it feasible to combine both now if I tie my download to a phase strictly earlier than compile. I will try to test it as mentioned above, but even with a toolchains.xml file, it does not seem to work (same error). As our migration to java 11 should happen soon though, I think my fix will hold up to this point.  

Thanks a lot for your help !
